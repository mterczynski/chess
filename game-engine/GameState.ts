export enum GameState {
    UNSTARTED = 'UNSTARTED',
    IN_PROGRESS = 'IN_PROGRESS',
    WHITE_WON = 'WHITE_WON',
    BLACK_WON = 'BLACK_WON',
    DRAW = 'DRAW',
}
