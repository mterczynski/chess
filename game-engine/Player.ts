export enum Player {
    BLACK = 'BLACK',
    WHITE = 'WHITE'
}
