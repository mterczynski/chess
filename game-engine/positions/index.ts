export * from './ChessFile';
export * from './Position';
export * from './Rank';
