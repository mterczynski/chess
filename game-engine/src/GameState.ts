export enum GameState {
    UNSTARTED = "UNSTARTED",
    IN_PROGRESS = "IN_PROGRESS",
    WHITE_WON = "WHITE_WON",
    BLACK_WON = "BLACK_WON",
    DRAW_BY_STALEMATE = "DRAW_BY_STALEMATE",
    DRAW_BY_REPETITION = "DRAW_BY_REPETITION",

    /** Not implemented */
    DRAW_BY_INSUFFICIENT_MATERIAL = "DRAW_BY_INSUFFICIENT_MATERIAL",
    /** Not implemented */
    DRAW_BY_50_MOVE_RULE = "DRAW_BY_50_MOVE_RULE",
    /** Not implemented */
    DRAW_BY_75_MOVE_RULE = "DRAW_BY_75_MOVE_RULE",
    /** Not implemented */
    DRAW_BY_AGREEMENT = "DRAW_BY_AGREEMENT",
}
