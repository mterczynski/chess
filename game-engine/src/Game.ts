import _ from "lodash";
import { AvailableMoveCalculator } from "./availableMovesCalculation";
import { CheckCalculator } from "./availableMovesCalculation/CheckCalculator";
import { Board } from "./Board";
import { createNewBoard } from "./utils";
import { GameState } from "./GameState";
import { isCastleablePiece, negatePlayer } from "./utils";
import { EnPassantMove, Move, SpecialMoveType } from "./Moves";
import { King, Piece, PieceType, Rook } from "./pieces";
import { Player } from "./Player";
import {
    addToFile,
    arePositionsEqual,
    ChessFile,
    getFileDifference,
} from "./positions";

export class Game {
    static readonly boardSize = 8;
    private state: GameState = GameState.UNSTARTED;
    private currentPlayer: Player = Player.WHITE;
    private board: Board = createNewBoard();
    private moves: Move[] = [];

    constructor(
        private readonly checkCalculator = new CheckCalculator(),
        private readonly availableMoveCalculator: AvailableMoveCalculator = new AvailableMoveCalculator(
            checkCalculator
        )
    ) {}

    getState(): GameState {
        return this.state;
    }

    /** Returns clone of the original board */
    getBoard(): Board {
        return _.cloneDeep(this.board);
    }

    move(move: Move): void {
        if (this.isGameOver()) {
            throw new Error(
                "Game has already finished, no more moves can be made"
            );
        }

        const updatedMove = this.checkMoveValidityAndAddMetadata(move);
        if (!updatedMove) {
            throw new Error("Invalid move");
        }

        this.applyMove(updatedMove);
        this.state = this.getNewGameStateAfterMove();

        if (this.state === GameState.IN_PROGRESS) {
            this.changePlayer();
        }
    }

    getAvailableMovesForPlayer(): Move[] {
        return this.availableMoveCalculator.getAvailableMovesForPlayer(
            this.board,
            this.currentPlayer,
            this.getLastMove()
        );
    }

    getCurrentPlayer(): Player | null {
        if (this.isGameOver()) {
            return null;
        }

        return this.currentPlayer;
    }

    clone(): Game {
        const clone = new Game();
        clone.state = this.state;
        clone.currentPlayer = this.currentPlayer;
        clone.board = _.cloneDeep(this.board);
        clone.moves = _.cloneDeep(this.moves);

        return clone;
    }

    private getLastMove(): Move | null {
        return this.moves.slice(-1)[0] || null;
    }

    private changePlayer(): void {
        this.currentPlayer =
            this.currentPlayer === Player.WHITE ? Player.BLACK : Player.WHITE;
    }

    private isGameOver(): boolean {
        const draws = [
            GameState.DRAW_BY_50_MOVE_RULE,
            GameState.DRAW_BY_75_MOVE_RULE,
            GameState.DRAW_BY_AGREEMENT,
            GameState.DRAW_BY_INSUFFICIENT_MATERIAL,
            GameState.DRAW_BY_REPETITION,
            GameState.DRAW_BY_STALEMATE,
        ];

        return [GameState.WHITE_WON, GameState.BLACK_WON, ...draws].includes(
            this.state
        );
    }

    private checkMoveValidityAndAddMetadata(move: Move): Move | null {
        const availableMoves = this.getAvailableMovesForPlayer();
        const availableMove = availableMoves.find(
            (availableMove) =>
                arePositionsEqual(availableMove.from, move.from) &&
                arePositionsEqual(availableMove.to, move.to)
        );

        return availableMove || null;
    }

    private castleRook(kingMove: Move) {
        const rank = kingMove.from.rank;
        const isKingMoveToTheRight =
            getFileDifference(kingMove.to.file, kingMove.from.file) < 0;
        const rookFile = isKingMoveToTheRight ? ChessFile.H : ChessFile.A;

        const rook = this.board[rookFile][rank] as Rook;
        const newRookFile = addToFile(
            kingMove.to.file,
            isKingMoveToTheRight ? -1 : 1
        ) as ChessFile;

        this.board[newRookFile][rank] = rook;
        this.board[rookFile][rank] = null;
    }

    private checkForCastle(move: Move, piece: Piece) {
        const isKing = piece.type === PieceType.KING;
        const filesToMove = Math.abs(
            getFileDifference(move.from.file, move.to.file)
        );
        return isKing && filesToMove === 2;
    }

    private enPassantIfAppliable(move: Move, piece: Piece) {
        if ((move as EnPassantMove).type === SpecialMoveType.EN_PASSANT) {
            this.board[move.to.file][
                move.to.rank +
                    (this.getCurrentPlayer() === Player.WHITE ? -1 : 1)
            ] = null;
        }
    }

    private applyMove(move: Move): void {
        const piece = this.board[move.from.file][move.from.rank] as Piece;
        this.board[move.to.file][move.to.rank] = piece;
        this.board[move.from.file][move.from.rank] = null;

        if (isCastleablePiece(piece)) {
            piece.hasMoved = true;

            // TODO - use SpecialMoveType (?)
            if (this.checkForCastle(move, piece)) {
                this.castleRook(move);
            }
        }

        this.enPassantIfAppliable(move, piece);
        this.moves.push(move);
    }

    // todo - extract to separate class (?)
    // maybe extract it to a class called GameStateHandler (?)
    private getNewGameStateAfterMove(): GameState {
        if (this.state === GameState.UNSTARTED) {
            return GameState.IN_PROGRESS;
        }

        const enemy = negatePlayer(this.currentPlayer);
        const enemyMoves =
            this.availableMoveCalculator.getAvailableMovesForPlayer(
                this.board,
                enemy,
                this.getLastMove()
            );

        if (enemyMoves.length === 0) {
            const availableCurrentPlayerMoves =
                this.availableMoveCalculator.getAvailableMovesForPlayer(
                    this.board,
                    this.currentPlayer,
                    this.getLastMove()
                );

            const checkingPiecesOfCurrentPlayer =
                this.checkCalculator.getCheckingEnemyPieces(
                    enemy,
                    this.board,
                    availableCurrentPlayerMoves
                );

            if (checkingPiecesOfCurrentPlayer.length === 0) {
                return GameState.DRAW_BY_STALEMATE;
            } else {
                return this.currentPlayer === Player.WHITE
                    ? GameState.WHITE_WON
                    : GameState.BLACK_WON;
            }
        }

        return this.state;
    }
}
