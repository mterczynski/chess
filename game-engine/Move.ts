import { PieceType } from "./pieces";
import { Position } from "./positions";

export interface Move {
    from: Position,
    to: Position,
}
