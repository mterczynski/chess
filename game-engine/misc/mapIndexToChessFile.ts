import { Game } from "..";
import { ChessFile } from "../positions";

export const mapIndexToChessFile = (index: number): ChessFile => {
    if(index < 0 || index > Game.boardSize - 1) {
        throw new Error('Invalid index');
    }

    return [...Object.values(ChessFile)][index];
}
