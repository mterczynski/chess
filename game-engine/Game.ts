import _ from "lodash";
import { AvailableMoveCalculator } from "./availableMovesCalculation";
import { Board } from "./Board";
import { isCastleablePiece } from "./castling";
import { GameState } from "./GameState";
import { Move } from "./Move";
import { Piece, PieceType, PieceWithPosition } from "./pieces";
import { PieceFactory } from "./pieces/PieceFactory";
import { Player } from "./Player";
import { ChessFile } from "./positions";

export class Game {
    static readonly boardSize = 8;
    private state: GameState = GameState.UNSTARTED;
    private currentPlayer: Player = Player.WHITE;
    private board: Board = this.createBoard();
    private lastMove: Move | null = null;

    constructor(
        private readonly pieceFactory: PieceFactory = new PieceFactory(),
        private readonly availableMoveCalculator: AvailableMoveCalculator = new AvailableMoveCalculator(),
    ) {}

    getState(): GameState {
        return this.state;
    }

    /** Returns clone of original board */
    getBoard(): Board {
        return _.cloneDeep(this.board);
    }

    move(move: Move): void {
        if(this.isValidMove(move)) {
            this.applyMove(move);
            this.changePlayer();
        } else {
            throw new Error('Invalid move');
        }
    }

    getAvailableMovesForPlayer(): Move[] {
        const currentPlayerPieces = this.getCurrentPlayerPieces();

        return currentPlayerPieces.flatMap(piece => this.availableMoveCalculator.getAvailableMovesForPiece(piece, this.board, this.currentPlayer, this.lastMove));
    }

    getCurrentPlayer(): Player | null {
        if(this.isGameOver()) {
            return null;
        }

        return this.currentPlayer;
    }

    private getCurrentPlayerPieces(): PieceWithPosition[] {
        const pieces = Object.entries(this.board).flatMap(([fileName, fileContent]) => {
            return fileContent.map((piece, rank) => (piece === null ? null : {
                ...piece,
                position: {
                    file: fileName as ChessFile,
                    rank,
                }
            })).filter((square): square is PieceWithPosition => !!square)
        });

        return pieces.filter(piece => piece!.player === this.currentPlayer);
    }

    private changePlayer(): void {
        if(this.currentPlayer === Player.WHITE) {
            this.currentPlayer = Player.BLACK;
        } else {
            this.currentPlayer = Player.WHITE;
        }
    }

    private isGameOver(): boolean {
        return this.state === GameState.WHITE_WON || this.state === GameState.BLACK_WON || this.state === GameState.DRAW
    }

    private isValidMove(move: Move): boolean {
        return this.hasMove(this.getAvailableMovesForPlayer(), move);
    }

    private hasMove(moves: Move[], move: Move): boolean {
        return moves
            .filter(_move => _move.from.file === move.from.file)
            .filter(_move => _move.from.rank === move.from.rank)
            .filter(_move => _move.to.file === move.to.file)
            .filter(_move => _move.to.rank === move.to.rank)
            .length >= 1;
    }

    private applyMove(move: Move): void {
        const piece = this.board[move.from.file][move.from.rank] as Piece;
        this.board[move.to.file][move.to.rank] = piece;
        this.board[move.from.file][move.from.rank] = null;

        if(isCastleablePiece(piece)) {
            piece.hasMoved = true;
        }
    }

    private createBoard(): Board {
        const createSymetricalFile = (pieceType: PieceType): (Piece | null)[] => {
            return [
                null,
                this.pieceFactory.createPiece(pieceType, Player.WHITE),
                this.pieceFactory.createPiece(PieceType.PAWN, Player.WHITE),
                null,
                null,
                null,
                null,
                this.pieceFactory.createPiece(PieceType.PAWN, Player.BLACK),
                this.pieceFactory.createPiece(pieceType, Player.BLACK)
            ]
        }

        return {
            [ChessFile.A]: createSymetricalFile(PieceType.ROOK),
            [ChessFile.B]: createSymetricalFile(PieceType.KNIGHT),
            [ChessFile.C]: createSymetricalFile(PieceType.BISHOP),
            [ChessFile.D]: createSymetricalFile(PieceType.QUEEN),
            [ChessFile.E]: createSymetricalFile(PieceType.KING),
            [ChessFile.F]: createSymetricalFile(PieceType.BISHOP),
            [ChessFile.G]: createSymetricalFile(PieceType.KNIGHT),
            [ChessFile.H]: createSymetricalFile(PieceType.ROOK),
        }
    }
}
