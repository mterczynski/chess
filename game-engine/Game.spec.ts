import { Game } from "./Game";
import { GameState } from "./GameState";
import { mapIndexToChessFile } from "./mapIndexToChessFile";
import { Move } from "./Move";
import { Player } from "./Player";
import { ChessFile } from "./positions";

describe('Game', () => {
    let game: Game;

    beforeEach(() => {
        game = new Game();
    })

    describe('getState', () => {
        it('returns GameState.UNSTARTED by default', () => {
            expect(game.getState()).toBe(GameState.UNSTARTED);
        });
    });

    describe('getCurrentPlayer', () => {
        it('returns Player.WHITE by default', () => {
            expect(game.getCurrentPlayer()).toBe(Player.WHITE);
        });

        it('returns Player.BLACK after making a move, then WHITE after making another move', () => {
            game.move({
                from: {
                    file: ChessFile.E,
                    rank: 2,
                },
                to: {
                    file: ChessFile.E,
                    rank: 4,
                },
            });
            expect(game.getCurrentPlayer()).toBe(Player.BLACK);
            game.move({
                from: {
                    file: ChessFile.E,
                    rank: 7,
                },
                to: {
                    file: ChessFile.E,
                    rank: 5,
                },
            });
            expect(game.getCurrentPlayer()).toBe(Player.WHITE);
        });

        it('returns null after game is finished', () => {
            // todo
        });
    });

    describe('move', () => {
        it('throws an error if invalid move was passed', () => {
            expect(() => {
                game.move({
                    from: {
                        file: ChessFile.A,
                        rank: 1,
                    },
                    to: {
                        file: ChessFile.A,
                        rank: 2,
                    },
                })
            }).toThrow('Invalid move');
        });
    });

    describe('getBoard', () => {
        it('returns starting board when called on new game', () => {
            const board = game.getBoard();

            expect(board).toMatchSnapshot();
        });

        it('returns correct board after first move', () => {
            const board = game.getBoard();

            game.move({
                from: {
                    file: ChessFile.E,
                    rank: 2
                },
                to: {
                    file: ChessFile.E,
                    rank: 4
                },
            })

            expect(board).toMatchSnapshot();
        });
    });

    // describe('getAvailableMovesForPiece', () => {
    //     it('should work for E pawn in new game', () => {
    //     });
    // })


    // describe('getAvailableMoves', () => {
    //     it('returns correct moves when called in new game', () => {
    //         const pawnMoves: Move[] = Array(Game.boardSize).fill(null).flatMap((_, index) => {
    //             const file = mapIndexToChessFile(index);
    //             return [
    //                 {from: {file: file, rank: 2}, to: {file: file, rank: 3}},
    //                 {from: {file: file, rank: 2}, to: {file: file, rank: 4}}
    //             ]
    //         });

    //         const knightMoves: Move[] = [
    //             {from: {file: ChessFile.B, rank: 1}, to: {file: ChessFile.A, rank: 3}},
    //             {from: {file: ChessFile.B, rank: 1}, to: {file: ChessFile.C, rank: 3}},
    //             {from: {file: ChessFile.G, rank: 1}, to: {file: ChessFile.F, rank: 3}},
    //             {from: {file: ChessFile.G, rank: 1}, to: {file: ChessFile.H, rank: 3}},
    //         ]

    //         const expectedMoves: Move[] = [
    //             ...pawnMoves,
    //             ...knightMoves,
    //         ];

    //         const result = game.getAvailableMovesForPlayer();

    //         expect(result).toEqual(expect.arrayContaining(expectedMoves));
    //         expect(result.length).toEqual(expectedMoves.length);
    //     });
    // });

});
