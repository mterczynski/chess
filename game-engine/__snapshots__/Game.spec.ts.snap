// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game getBoard returns correct board after first move 1`] = `
Object {
  "A": Array [
    null,
    Object {
      "hasMoved": false,
      "player": "WHITE",
      "type": "R",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "hasMoved": false,
      "player": "BLACK",
      "type": "R",
    },
  ],
  "B": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "N",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "N",
    },
  ],
  "C": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "B",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "B",
    },
  ],
  "D": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "Q",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "Q",
    },
  ],
  "E": Array [
    null,
    Object {
      "hasMoved": false,
      "player": "WHITE",
      "type": "K",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "hasMoved": false,
      "player": "BLACK",
      "type": "K",
    },
  ],
  "F": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "B",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "B",
    },
  ],
  "G": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "N",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "N",
    },
  ],
  "H": Array [
    null,
    Object {
      "hasMoved": false,
      "player": "WHITE",
      "type": "R",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "hasMoved": false,
      "player": "BLACK",
      "type": "R",
    },
  ],
}
`;

exports[`Game getBoard returns starting board when called on new game 1`] = `
Object {
  "A": Array [
    null,
    Object {
      "hasMoved": false,
      "player": "WHITE",
      "type": "R",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "hasMoved": false,
      "player": "BLACK",
      "type": "R",
    },
  ],
  "B": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "N",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "N",
    },
  ],
  "C": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "B",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "B",
    },
  ],
  "D": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "Q",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "Q",
    },
  ],
  "E": Array [
    null,
    Object {
      "hasMoved": false,
      "player": "WHITE",
      "type": "K",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "hasMoved": false,
      "player": "BLACK",
      "type": "K",
    },
  ],
  "F": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "B",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "B",
    },
  ],
  "G": Array [
    null,
    Object {
      "player": "WHITE",
      "type": "N",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "player": "BLACK",
      "type": "N",
    },
  ],
  "H": Array [
    null,
    Object {
      "hasMoved": false,
      "player": "WHITE",
      "type": "R",
    },
    Object {
      "player": "WHITE",
      "type": "P",
    },
    null,
    null,
    null,
    null,
    Object {
      "player": "BLACK",
      "type": "P",
    },
    Object {
      "hasMoved": false,
      "player": "BLACK",
      "type": "R",
    },
  ],
}
`;
